sphinx/ext/asyncio.py

```diff
@@ -0,0 +1,104 @@
+# -*- coding: utf-8 -*-
+"""
+    sphinx.ext.asyncio
+    ~~~~~~~~~~~~~~~~~~
+
+    Adds directives for :py:class:`asyncio`.
+
+    .. rst:directive:: .. py:coroutine:: name(parameters)
+
+       Like :rst:dir:`py:function`, but for :py:class:`asyncio.coroutine`.
+
+       .. versionadded:: 1.4
+
+    .. rst:directive:: .. py:coroutinemethod:: name(parameters)
+
+       Like :rst:dir:`py:method`, but for :py:class:`asyncio.coroutine`.
+
+       .. versionadded:: 1.4
+
+    :copyright: Copyright 2007-2015 by the Sphinx team, see AUTHORS.
+    :license: BSD, see LICENSE for details.
+"""
+
+import logging
+from sphinx import addnodes
+from sphinx.domains.python import PyModulelevel, PyClassmember
+from sphinx.ext.autodoc import FunctionDocumenter, MethodDocumenter
+
+try:
+    from asyncio import iscoroutinefunction
+    supported = True
+except ImportError:
+    try:
+        from trollius import iscoroutinefunction
+        supported = True
+    except ImportError:
+        supported = False
+
+        def iscoroutinefunction(func):
```

Comment:
The [implementation in CPython](https://hg.python.org/cpython/file/393189326adb/Lib/asyncio/coroutines.py#l164) looks easy enough.  Could we just copy it to `sphinx/util/pycompat.py` (or `inspect.py`, which you've used for `isgeneratorfunction` already?)